### DISCUSSION TITLE
QC: Document `{cli}` usage

### DISCUSSION THREAD

**Discussion by mattroumaya** [Gilead-Public/gsm#503] [General]
## QC Details
Create documentation for QC and for creating new `*_Map` and `*_Assess` functions, where `{cli}` alerts should be consistent.

**Comment by mattroumaya** [comment]
After some investigation, it seems like a somewhat considerable refactor to capture the dynamic `{cli}` messages to use in logging/reporting. 

I avoided using `sink` because I feel that it would introduce some potential risk/messyness for the end-user, and not totally sure how it would work across multiple functions while running assessments (i.e. `Study_Assess()`)

A first iteration and possible implementation could be to use a `cli_wrapper()` utility function: 
```
cli_wrapper <- function(str, type, bQuiet = TRUE, return_message=TRUE, ...){

  str <- with(list(...), glue::glue(str))
  if(return_message){
    msg <- glue::glue('cli::cli_{type}("{str}")')
    eval(parse(text=msg))
  }
  if(return_message){
    invisible(str)
  }
}
```

This could return something like below. Note that the info about imputing 0s is from `MergeSubjects`, where I haven't implemented the `cli_wrapper`, so that messaging is missing from this example.
```r
> ae_mapped <- AE_Map_Raw(bQuiet = F, bReturnChecks = T)

── Checking Input Data for `AE_Map_Raw()` ──

✔ No issues found for `AE_Map_Raw()`

── Initializing `AE_Map_Raw` ──

! 283 ID(s) in subject data not found in domain data.
These participants will have 0s imputed for the following domain data columns: Count. 
NA's will be imputed for all other columns.
✔ `AE_Map_Raw` returned output with 1297 rows.


# when bReturnChecks == TRUE:
> ae_mapped$message
[1] "Initializing `AE_Map_Raw`"                    "`AE_Map_Raw` returned output with 1297 rows."

```

And below would be the refactor needed to any functions using `{cli}`:
```r
AE_Map_Raw <- function(
    dfs = list(
      dfAE = clindata::rawplus_ae,
      dfSUBJ = clindata::rawplus_subj
    ),
    lMapping = clindata::mapping_rawplus,
    bReturnChecks = FALSE,
    bQuiet = TRUE
) {
  checks <- gsm::CheckInputs(
    context = "AE_Map_Raw",
    dfs = dfs,
    bQuiet = bQuiet,
    mapping = lMapping
  )
  message <- NULL
  # Run mapping if checks passed.
  if (checks$status) {
    message <- c(message,
                 cli_wrapper(str = "Initializing `AE_Map_Raw`",
                             type = "h2",
                             bQuiet = bQuiet)
                 )


    # Standarize Column Names
    dfAE_mapped <- dfs$dfAE %>%
      select(SubjectID = lMapping[["dfAE"]][["strIDCol"]])

    dfSUBJ_mapped <- dfs$dfSUBJ %>%
      select(
        SubjectID = lMapping[["dfSUBJ"]][["strIDCol"]],
        SiteID = lMapping[["dfSUBJ"]][["strSiteCol"]],
        Exposure = lMapping[["dfSUBJ"]][["strTimeOnTreatmentCol"]]
      )

    # Create Subject Level AE Counts and merge dfSUBJ
    dfInput <- dfAE_mapped %>%
      group_by(.data$SubjectID) %>%
      summarize(Count = n()) %>%
      ungroup() %>%
      gsm::MergeSubjects(dfSUBJ_mapped, vFillZero = "Count", bQuiet = bQuiet) %>%
      mutate(Rate = .data$Count / .data$Exposure) %>%
      select(.data$SubjectID, .data$SiteID, .data$Count, .data$Exposure, .data$Rate)
    message <- c(message, cli_wrapper(str = "`AE_Map_Raw` returned output with {nrow(dfInput)} rows.",
                                      type = "alert_success",
                                      bQuiet = bQuiet,
                                      dfInput = dfInput))
  } else {
    message <- c(message,
                 cli_wrapper(str = "`AE_Map_Raw` did not run because of failed check.",
                             type = "alert_warning",
                             bQuiet = bQuiet))
    dfInput <- NULL
  }

  if (bReturnChecks) {
    return(list(df = dfInput, lChecks = checks, message = message))
  } else {
    return(dfInput)
  }
}
```

**Comment by jwildfire** [comment]
@samussiah @jwildfire @kodesiba to review before Monday's meeting.

**Comment by kodesiba** [comment]
I think I have seen it mostly used with wrappers in other packages.  One of the industry packages is Merck's `{pkglite}` that uses similar [utility functions](https://github.com/Merck/pkglite/blob/7dbcd599d7fc7dcb6c26e0023fc4cefb80bd0c59/R/utils-cli.R) to send different messages to the log using `{cli}`.  I do really like the implementation here for a wrapper as it means you can pass the type and other pieces to it.  If we wanted to there's also always [`{crayon}`](https://cran.r-project.org/web/packages/crayon/index.html) that can color things and make different pieces stand out to the user.

**Comment by samussiah** [comment]
It's verbose but that's the point I suppose.  The goal here then is to optionally present select `cli` output in the report?

I do like the wrapper because it's modular and easier to read than `if (!bQuiet) cli_...`.  It serves the purpose.

It would be nice to be able to capture the `cli`-formatted text but the `cli` functions seem to only return an ID.
